### PR TITLE
Move out consensus logic from ToC

### PR DESCRIPTION
**Pull request by e-ivkov**
Relates to #639

### Changes
The PR moves out consensus logic from ToC into `Dispatcher` and `ConsensusState`. Also several places in the state logic were simplified and moved out into corresponding structs.

### Further work
There are still some rough edges that I want to address in later smaller PRs:
1. Improve initialization in main - right now it is just too complex
2. Remove `is_new_deployment` state
3. Move out from `consensus_state.rs` into several files in a module

### CODE REVIEW

File: tests/consensus_tests/utils.py [hunk 1]
```diff
@@ -155,6 +155,8 @@ def wait_for_leader_setup(peer_api_uri: str):
             raise Exception(f"Leader was not established in time ({WAIT_TIME_SEC} sec)")
         else:
             time.sleep(RETRY_INTERVAL_SEC)
+    # It seems that waiting for leader to be recognised is not enough
+    time.sleep(3)
```
Comment (generall) [inline tests/consensus_tests/utils.py:159]: Should we maybe create an issue for this?

Comment (e-ivkov) [reply tests/consensus_tests/utils.py:159]: I believe it's related to the one that Arnaud is working on currently. If we can add multiple peers concurrently without a problem, then we can remove the wait here.

Comment (generall) [reply tests/consensus_tests/utils.py:159]: https://github.com/qdrant/qdrant/issues/702 pls add some more context to this issue @e-ivkov

**Review by generall — APPROVED** [review]

**Review by agourlay — APPROVED** [review]
Thank you for this refactoring 👍 

I just have added a small question for my curiosity.

File: lib/collection/src/lib.rs [hunk 2]
```diff
@@ -217,17 +217,17 @@ impl CollectionShardDistribution {
 
 #[derive(Clone)]
 pub struct ChannelService {
-    ip_to_address: Arc<std::sync::RwLock<HashMap<u64, Uri>>>,
-    channel_pool: Arc<TransportChannelPool>,
+    pub id_to_address: Arc<parking_lot::RwLock<HashMap<u64, Uri>>>,
```
[the remark this answers is not in the document]
Comment (e-ivkov) [reply lib/collection/src/lib.rs:220]: We have a deadlock detection possibility for parking lot. Also it ignores poisoning - so we don't have so many unwraps or error conversions. And we were not using lock poisoning (the thing it is useful for) for shutting down the whole app anyway.

**Pull request merged**